tls/s2n_cipher_suites.c

```diff
@@ -1168,13 +1216,175 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
     return 0;
 }
 
+/**
+ * @brief While a potential cipher suite match has been identified, we still need to check if the cipher suite
+ *        can actually be used for the remainder of the connection. These checks ensure that there exists an
+ *        implementation, whether the versions and PSKs are compatible.
+ * 
+ * @param conn s2n_connection context
+ * @param potential_match Potential cipher suite match
+ * @return true Cipher suite match is usable
+ * @return false Cipher suite match is not usable for this connection
+ */
+static bool s2n_cipher_suite_match_is_valid(struct s2n_connection* conn, struct s2n_cipher_suite* potential_match) {
```

Comment:
I think it's nice to document internal functions to make it easier to understand what they do.

Reply:
I do think comments are good here, it just seems like a lot of work to do this comment style everywhere in the code base.

Reply:
1. Public api or no, is there a con to commenting in this style? I think this is makes the function clearer for developers
2. Is there a benefit to setting an output parameter instead of just returning the value instead?

Reply:
The formatting is nice for public APIs because it's parsed by tools and turned into very human readable text on a UI. These comments aren't parsed by a tool, and the `@blah name comment` format isn't necessarily the most human readable.

Reply:
The purpose of doxygen and internal comments is also very different. Doxygen comments assume the reader is just looking at the function definition with the intention of using it. Internal comments assume the reader is looking at the implementation with the intention of modifying it. You should focus on "why" most of the time, and only explain "what" for unavoidably complex functions.

Reply:
And just to address question 2: If a function can error than it needs to have a return value of S2N_RESULT. And this function can error, because you call POSIX_GUARD_RESULT. Otherwise it's fine to return a bool.

Reply:
I still think this comment structure communicates the 'why' and 'what' better than a random multi-line comment. As for readability, I don't see `@blah name content` as not human readable. Regardless, I'll just collapse this comment into one without this formatting.

Reply:
One benefit to using `S2N_RESULT` over `bool` is you get the line on which the check failed rather than `true`/`false` at the very end.

We also have all of the safety macros which make this kind of logic much more declarative; where you're asserting what the state should be rather than describing control flow.

```c
    RESULT_ENSURE_EQ(conn->actual_protocol_version >= S2N_TLS13, potential_match->minimum_required_tls_version >= S2N_TLS13);
    RESULT_ENSURE(potential_match->available);
    RESULT_GUARD_POSIX(s2n_is_cipher_suite_valid_for_auth(conn, potential_match));
```